### PR TITLE
model: Introduce and use a generic findFilesWithKnownExtensions() function

### PR DESCRIPTION
**Pull request by sschuberth**
Signed-off-by: Sebastian Schuberth <sebastian.schuberth@bosch.io>

### CODE REVIEW

File: model/src/main/kotlin/FileFormat.kt [hunk 1]
```diff
@@ -64,6 +64,16 @@ enum class FileFormat(val mapper: ObjectMapper, val fileExtension: String, varar
...
          */
         fun forFile(file: File): FileFormat = forExtension(file.extension)
+
+        /**
+         * Return a list of all files inside [directory] with known extensions that can be deserialized.
+         */
+        fun findDeserializableFiles(directory: File): List<File> {
+            val allFileExtensions = enumValues<FileFormat>().flatMap { it.fileExtensions }
+            return directory.walkBottomUp().filter {
+                it.isFile && it.extension in allFileExtensions && !it.isHidden
```
Comment (sschuberth) [inline model/src/main/kotlin/FileFormat.kt:74]: @fviernau do you guys actually make use of the hidden flag to "disable" files, or could we drop `!it.isHidden` for simplicity?

Comment (fviernau) [reply model/src/main/kotlin/FileFormat.kt:74]: We are not using the "hidden" flag to "disable" package configurations.

File: model/src/main/kotlin/FileFormat.kt [hunk 2]
```diff
@@ -64,6 +64,16 @@ enum class FileFormat(val mapper: ObjectMapper, val fileExtension: String, varar
          * Return the [FileFormat] for the given [file], or `null` if there is none.
          */
         fun forFile(file: File): FileFormat = forExtension(file.extension)
+
```
Comment (fviernau) [inline model/src/main/kotlin/FileFormat.kt:67]: commit: What's the reason for extracting this function?

**Review by fviernau — CHANGES_REQUESTED** [review]

File: model/src/main/kotlin/FileFormat.kt [hunk 3]
```diff
@@ -64,6 +64,16 @@ enum class FileFormat(val mapper: ObjectMapper, val fileExtension: String, varar
...
+        /**
+         * Return a list of all files inside [directory] with known extensions that can be deserialized.
+         */
+        fun findDeserializableFiles(directory: File): List<File> {
```
Comment (fviernau) [inline model/src/main/kotlin/FileFormat.kt:71]: Would it make sense to name this e.g. `findFilesWithKnownExtension`
to not communicate that the returned files can actually be deserialized?

File: model/src/main/kotlin/FileFormat.kt (hunk 2 above)
Comment (fviernau) [reply model/src/main/kotlin/FileFormat.kt:67]: I'm not certain how we do it for `ort.yml` and other config files.
However, would it make sense to limit the support to just `yml`?

Comment (sschuberth) [reply model/src/main/kotlin/FileFormat.kt:67]: We're planning to make reuse of it in an upcoming PR for other YAML / JSON files in a directory tree.

File: model/src/main/kotlin/FileFormat.kt (hunk 3 above)
Comment (sschuberth) [reply model/src/main/kotlin/FileFormat.kt:71]: Makes sense, will do.

File: model/src/main/kotlin/FileFormat.kt (hunk 2 above)
Comment (sschuberth) [reply model/src/main/kotlin/FileFormat.kt:67]: > However, would it make sense to limit the support to just `yml`?

I don't think so. While unusual and not really documented, we actually would support `json` files anywhere where we usually use `yml` files, so I'd like to keep the function generic.

**Review by MarcelBochtler — APPROVED** [review]

**Review by fviernau — DISMISSED** [review]

**Review by fviernau — APPROVED** [review]

**Review by MarcelBochtler — APPROVED** [review]

**Pull request merged**